### DISCUSSION TITLE
posts/0day-chp03/

### DISCUSSION THREAD

**Discussion by giscus** [brant-ruan/fernweh-comments#23] [Announcements]
# posts/0day-chp03/

0day安全｜03 开发shellcode的艺术

https://blog.wohin.me/posts/0day-chp03/

**Comment by zha0cai** [comment]
请教一下，我调试最后的 bind_shell 在 
;initialize winsock
    push esp  ;use stack for WSADATA 
	push 0x02 ;wVersionRequested 
	lodsd 
	call eax  ;WSAStartup // 这里会出错，eax 此时是 5F327377 即 "ws2_32" 会导致出错

**Reply by zha0cai** [reply]
ok，直接用机器码倒是可以解决，是因为 shellcode 其实地址为 eax 不能直接运行，是需要加载的方式吧。
int main()
{
	__asm{ ... }
}

**Reply by brant-ruan** [reply]
这个实验对我来说有点久远了。看了一下，你要确保`lodsd`加载到`eax`的地址是正确的；进一步地，`esi`指向的地方是否有配置为正确的函数地址。可以按照这个思路回溯debug一下。

**Reply by zha0cai** [reply]
thx
